internal/contour/handler.go

```diff
@@ -164,11 +181,37 @@ func (e *EventHandler) Start(ctx context.Context) error {
 				// not to process it.
 				e.incSequence()
 			}
+
+			// We're done processing this event
+			if updateOpAdd, ok := op.(opAdd); ok {
+				if updateOpAdd.isInInitialList {
+					e.syncTracker.Finished()
+				}
+			}
 		case <-pending:
+			// Ensure informer caches are synced.
+			// Schedule a retry for dag rebuild if cache is not synced yet.
+			// Note that we can't block and wait for the cache sync as it depends on progress of this loop.
+			if !e.syncTracker.HasSynced() {
+				e.Info("skipping dag rebuild as cache is not synced")
+				timer.Reset(e.holdoffDelay)
+				break
+			}
+
 			e.WithField("last_update", time.Since(lastDAGRebuild)).WithField("outstanding", reset()).Info("performing delayed update")
-			e.rebuildDAG()
-			e.incSequence()
+
+			// Rebuild the DAG
+			latestDAG := e.builder.Build()
 			lastDAGRebuild = time.Now()
+			e.initialDagBuilt = true
+
+			// Send update to the Observer and StatusUpdater
+			e.observer.OnChange(latestDAG)
+			for _, upd := range latestDAG.StatusCache.GetStatusUpdates() {
+				e.statusUpdater.Send(upd)
+			}
+
+			e.incSequence()
```

Comment:
One more thing, I think the order should be like this:

```go
// Build a new DAG and sends it to the Observer.
latestDAG := e.builder.Build()
e.observer.OnChange(latestDAG)

// Allow XDS server to start (if it hasn't already).
e.initialDagBuilt = true

// Update the status on objects.
for _, upd := range latestDAG.StatusCache.GetStatusUpdates() {
    e.statusUpdater.Send(upd)
}
```

That is, the flag is set to `true` after letting the observers run, but before status updater runs. There the observer part will populate the xdscaches with Envoy config. This is then input for Contour's XDS server to serve to Envoys. 

I'd also appreciate explicit comment above `e.initialDagBuilt = true` to highlight it has rather important side effect of starting the XDS server.